finality-aleph/src/substrate_network.rs

```diff
@@ -0,0 +1,413 @@
+        NetworkService::request_justification(self, hash, number)
+    }
+
+    fn request_stale_block(&self, hash: B::Hash, number: NumberFor<B>) {
+        // The below comment is adapted from substrate:
+        // Notifies the sync service to try and sync the given block from the given peers. If the given vector
+        // of peers is empty (as in our case) then the underlying implementation should make a best effort to fetch
+        // the block from any peers it is connected to.
+        NetworkService::set_sync_fork_request(self, Vec::new(), hash, number)
+    }
+
+    /// Clear all pending justification requests.
+    fn clear_justification_requests(&self) {
+        NetworkService::clear_justification_requests(self)
+    }
+}
+
+#[derive(PartialEq, Eq, Copy, Clone, Debug, Hash)]
+pub struct PeerId(SubstratePeerId);
+
+impl From<PeerId> for SubstratePeerId {
+    fn from(wrapper: PeerId) -> Self {
+        wrapper.0
+    }
+}
+
+impl From<SubstratePeerId> for PeerId {
+    fn from(id: SubstratePeerId) -> Self {
+        PeerId(id)
+    }
+}
+
+impl Encode for PeerId {
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.0.to_bytes().using_encoded(f)
+    }
+}
+
+impl Decode for PeerId {
+    fn decode<I: codec::Input>(value: &mut I) -> Result<Self, codec::Error> {
+        let bytes = Vec::<u8>::decode(value)?;
+        SubstratePeerId::from_bytes(&bytes)
+            .map_err(|_| "PeerId not encoded with to_bytes".into())
+            .map(|pid| pid.into())
+    }
+}
+
+impl PeerIdT for PeerId {}
+
+fn peer_id(protocol: &MultiaddressProtocol<'_>) -> Option<PeerId> {
+    match protocol {
+        MultiaddressProtocol::P2p(hashed_peer_id) => {
+            SubstratePeerId::from_multihash(*hashed_peer_id)
+                .ok()
+                .map(PeerId)
+        }
+        _ => None,
+    }
+}
+
+/// A wrapper for the Substrate multiaddress to allow encoding & decoding.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Multiaddress(Multiaddr);
+
+impl From<Multiaddr> for Multiaddress {
+    fn from(addr: Multiaddr) -> Self {
+        Multiaddress(addr)
+    }
+}
+
+impl From<Multiaddress> for Multiaddr {
+    fn from(addr: Multiaddress) -> Self {
+        addr.0
+    }
+}
+
+impl Encode for Multiaddress {
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.0.to_vec().using_encoded(f)
+    }
+}
+
+impl Decode for Multiaddress {
+    fn decode<I: codec::Input>(value: &mut I) -> Result<Self, codec::Error> {
+        let bytes = Vec::<u8>::decode(value)?;
+        Multiaddr::try_from(bytes)
+            .map_err(|_| "Multiaddr not encoded as bytes".into())
+            .map(|multiaddr| multiaddr.into())
+    }
+}
+
+impl MultiaddressT for Multiaddress {
+    type PeerId = PeerId;
+
+    fn get_peer_id(&self) -> Option<Self::PeerId> {
+        self.0
+            .iter()
+            .try_fold(None, |maybe_peer_id, protocol| match peer_id(&protocol) {
```

Comment:
use `fold` if possible ?

Reply:
Just as in the other case extracting the function removed the `try_`.